### PR TITLE
Update AFMoE architecture to use v5-style MoE impl

### PR DESCRIPTION
**Pull request by AutumnAurelium**
# What does this PR do?

This brings the Arcee AFMoE architecture in line with other MoE models' implementation patterns since v5. It also adds integration testing using Trinity Nano.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @Cyrilvallez

### CODE REVIEW

File: src/transformers/models/afmoe/modeling_afmoe.py [hunk 1]
```diff
@@ -254,15 +238,15 @@ def forward(self, hidden_states):
...
+        router_logits, top_scores, selected_experts = self.router(hidden_states, self.expert_bias)
 
         # Process through shared experts
         shared_output = self.shared_experts(hidden_states_flat).view(batch_size, seq_len, hidden_dim)
-        routed_output = self.experts(hidden_states, selected_experts, top_scores)
-        return shared_output + routed_output
+        routed_output = self.experts(hidden_states_flat, selected_experts, top_scores).view(
+            batch_size, seq_len, hidden_dim
+        )
+        return shared_output + routed_output, router_logits
```
Comment (ArthurZucker) [inline src/transformers/models/afmoe/modeling_afmoe.py]: router logits are not collected from this class you don't have to return them here and that allows for less changes in the forward!

File: src/transformers/models/afmoe/modular_afmoe.py [hunk 2]
```diff
@@ -311,7 +292,10 @@ def forward(
         # FFN with dual normalization
         residual = hidden_states
         hidden_states = self.pre_mlp_layernorm(hidden_states)
-        hidden_states = self.mlp(hidden_states)
+        if self.moe_enabled:
```
Comment (ArthurZucker) [inline src/transformers/models/afmoe/modular_afmoe.py]: not needed!

**Review by ArthurZucker — COMMENTED** [review]
sounds good thanks for updating!

**Comment by AutumnAurelium** [comment]
@ArthurZucker @Cyrilvallez 

Any update on getting this merged? Fixed problems mentioned above.

**Comment by winglian** [comment]
run-slow: afmoe

**Review by winglian — APPROVED** [review]
lgtm!

**Comment by winglian** [comment]
confirmed that model trains in axolotl, as well as loads experts as expected;
```
>>> import torch
>>> from transformers import AutoTokenizer, AutoModelForCausalLM
>>> model_id = "arcee-ai/Trinity-Nano-Base"
>>> tokenizer = AutoTokenizer.from_pretrained(model_id)
>>> model = AutoModelForCausalLM.from_pretrained(model_id, dtype=torch.float16, device_map="auto")
Loading weights: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1003/1003 [00:01<00:00, 948.02it/s, Materializing param=model.norm.weight]
>>> messages = [{"role":"user","content":"tell me about the culinary holy trinity"}]
>>> inputs = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt").to(model.device)
>>> outputs = model.generate(**inputs, max_new_tokens=400, temperature=0.6, top_p=0.95, top_k=20, do_sample=True)
>>> print(tokenizer.decode(outputs[0], skip_special_tokens=True))
user
tell me about the culinary holy trinity
assistant
The culinary holy trinity refers to a combination of three key ingredients that are essential in many dishes. The trinity varies depending on the cuisine, but a common trinity includes onions, bell peppers, and celery. This trinity is used in dishes like Cajun and Creole cuisine, where it forms the base for many sauces and stews.
>>> model.model.layers[2]
AfmoeDecoderLayer(
	(self_attn): AfmoeAttention(
		(q_proj): Linear(in_features=1024, out_features=1024, bias=False)
		(k_proj): Linear(in_features=1024, out_features=256, bias=False)
		(v_proj): Linear(in_features=1024, out_features=256, bias=False)
		(o_proj): Linear(in_features=1024, out_features=1024, bias=False)
		(q_norm): AfmoeRMSNorm((128,), eps=1e-05)
		(k_norm): AfmoeRMSNorm((128,), eps=1e-05)
		(gate_proj): Linear(in_features=1024, out_features=1024, bias=False)
		(rotary_fn): Func()
	)
	(input_layernorm): AfmoeRMSNorm((1024,), eps=1e-05)
	(post_attention_layernorm): AfmoeRMSNorm((1024,), eps=1e-05)
	(pre_mlp_layernorm): AfmoeRMSNorm((1024,), eps=1e-05)
	(post_mlp_layernorm): AfmoeRMSNorm((1024,), eps=1e-05)
	(mlp): AfmoeMoE(
		(router): AfmoeTokenChoiceRouter(
			(gate): Linear(in_features=1024, out_features=128, bias=False)
		)
		(shared_experts): AfmoeMLP(
			(gate_proj): Linear(in_features=1024, out_features=256, bias=False)
			(up_proj): Linear(in_features=1024, out_features=256, bias=False)
			(down_proj): Linear(in_features=256, out_features=1024, bias=False)
			(act_fn): SiLUActivation()
		)
		(experts): AfmoeExperts(
			(act_fn): SiLUActivation()
		)
	)
)
```

File: src/transformers/models/afmoe/modular_afmoe.py [hunk 3]
```diff
@@ -75,76 +77,55 @@ def forward(self, hidden_states: torch.Tensor, expert_bias: torch.Tensor):
...
 
-        weighted_outputs = (expert_outputs.to(torch.float32) * routing_weights.unsqueeze(-1)).to(hidden_states.dtype)
-        aggregated = torch.zeros_like(hidden_states_flat)
-        scatter_indices = token_indices.unsqueeze(-1).expand_as(weighted_outputs)
-        aggregated.scatter_add_(0, scatter_indices, weighted_outputs)
-        return aggregated.view(batch_size, seq_len, hidden_dim)
+        return final_hidden_states
 
 
 class AfmoeMoE(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/afmoe/modular_afmoe.py]: pretty sure you can now inherit this from another class! can you try ? 🤗

**Review by ArthurZucker — COMMENTED** [review]
LGTM just let's leverage modular in that case the MOE is standard can be inherited!

File: src/transformers/models/afmoe/modular_afmoe.py (hunk 3 above)
Comment (winglian) [reply src/transformers/models/afmoe/modular_afmoe.py]: @ArthurZucker Should this also be named `AfmoeSparseMoeBlock` for consistency?

File: tests/models/afmoe/test_modeling_afmoe.py [hunk 4]
```diff
@@ -120,4 +125,79 @@ def test_model_rope_scaling_frequencies(self):
...
+            max_new_tokens=num_tokens_to_generate,
+            do_sample=False,
+            cache_implementation="static",
+        )
+        static_text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
+        self.assertEqual(dynamic_text, static_text)
+
+        # Clear any existing cache object before static+compiled generation.
+        if hasattr(model, "_cache"):
+            del model._cache
```
Comment (ArthurZucker) [inline tests/models/afmoe/test_modeling_afmoe.py]: there's no `_cache` objecti in transformers

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/afmoe/modular_afmoe.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/afmoe/modular_afmoe.py]: yep perfect
